magma/spec/update_spec.rb

```diff
@@ -1262,6 +1258,234 @@ def update(revisions, user_type=:editor, params: {})
+        expect(Labors::Victim.first.monster_id).to eq(Labors::Monster.first.id)
+        expect(Labors::Monster.first.name).to eq(monster_identifier.identifier)
+        expect(Labors::Monster.first.labor_id).to eq(Labors::Labor.first.id)
+        expect(Labors::Labor.first.name).to eq(labor_identifier.identifier)
+        expect(Labors::Labor.first.project_id).to eq(Labors::Project.first.id)
+        expect(Labors::Project.first.name).to eq(project_identifier.identifier)
+
+      end
+
+    end
+
+    context 'pattern mode' do
+
+      it 'creates a record if the identifier matches the pattern in the gnomon grammar' do
+
+        Magma.instance.db[:flags].insert(
+          project_name: "labors",
+          flag_name: Magma::Flags::GNOMON_MODE[:name],
+          value: Magma::Flags::GNOMON_MODE[:pattern],
+          created_at: Time.now,
+          updated_at: Time.now,
+          )
+
+        grammar = create(:grammar, { project_name: 'labors', version_number: 1, config: VALID_GRAMMAR_CONFIG, comment: 'update' })
+        identifier = create_identifier("LABORS-LION-H2-C1", rule: 'victim', grammar: grammar)
+
+        update(
+          monster: {
+            identifier.identifier => { species: 'lion' }
+          }
+        )
+
+        expect(last_response.status).to eq(200)
+        expect(Labors::Monster.first.name).to eq(identifier.identifier)
+        expect(Labors::Monster.first.species).to eq('lion')
+      end
+
+      it 'updates a record if the identifier matches the pattern in the gnomon grammar' do
+
+        Magma.instance.db[:flags].insert(
+          project_name: "labors",
+          flag_name: Magma::Flags::GNOMON_MODE[:name],
+          value: Magma::Flags::GNOMON_MODE[:pattern],
+          created_at: Time.now,
+          updated_at: Time.now,
+          )
+
+        grammar = create(:grammar, { project_name: 'labors', version_number: 1, config: VALID_GRAMMAR_CONFIG, comment: 'update' })
+        identifier = create_identifier("LABORS-LION-H2-C1", rule: 'victim', grammar: grammar)
+
+        update(
+          monster: {
+            identifier.identifier => { species: 'lion' }
+          }
+        )
+
+        update(
+          monster: {
+            identifier.identifier => { species: 'tiger' }
+          }
+        )
+
+        expect(last_response.status).to eq(200)
+        expect(Labors::Monster.first.species).to eq('tiger')
+      end
+
+      it 'rejects a record if the identifier does not match the pattern in the gnomon grammar' do
+
+        Magma.instance.db[:flags].insert(
+          project_name: "labors",
+          flag_name: Magma::Flags::GNOMON_MODE[:name],
+          value: Magma::Flags::GNOMON_MODE[:pattern],
+          created_at: Time.now,
+          updated_at: Time.now,
+          )
+
+        grammar = create(:grammar, { project_name: 'labors', version_number: 1, config: VALID_GRAMMAR_CONFIG, comment: 'update' })
+        # TODO: why does this pass? This is an invalid identifier
```

Comment:
@graft is there a bug with Gnomon here?

Here this test creates an identifier using an invalid separator: `*`.

When it is validated via:

```ruby
gnomon_validation = Magma::Gnomon::Validation.new( Magma::Gnomon::Grammar::Parser.new(grammar[:config]) )
return if gnomon_validation.valid?
```
it somehow passes?

Reply:
Hmm, I think this is right? You inserted an invalid identifier into the db. Gnomon will complain if you try to register an invalid id via api, but once it is in the db, magma will treat it as ok without further validation.

Reply:
a bit more, when an identifier is created in Gnomon it notes the version of the grammar used to create the identifier (via foreign key to the grammar table i think); however most endpoints like decompose currently deal only with the current version of the grammar, so they might not decompose an old identifier correctly. It is pretty hard to do this well since an old grammars may deal in old models, which are no longer valid.

Reply:
After some thought, seems like we can't auto-attach old identifiers this way, since we can't necessarily decompose them or even line them up against existing models. A few options:

1. Allow the no-longer-valid identifier to be used in the update, but don't auto-attach the record using Gnomon decomposition in this case. The user can either explicitly attach this (like specify parents in the update), or leave it detached.
2. Optimistically try to decompose using the identifier's version of the grammar. This could produce a valid set of rules allowing attachment against the current set of models, in which case, why not attach? If they don't, the current method will skip anyway.
3. Raise an error for out-of-date identifiers and refuse to enter them.

Reply:
> Gnomon will complain if you try to register an invalid id via api, but once it is in the db, magma will treat it as ok without further validation.

Are there instances where we would not use the API to register an identifier?

Reply:
>Are there instances where we would not use the API to register an identifier?
Doubtful, but since the grammar may change over time, we will undoubtedly accumulate no-longer-valid identifiers anyway.

Reply:
For now, maybe let's raise an error for out-of-date identifiers. 

Conceptually the simplest and prevents potential data integrity issues. If it becomes cumbersome to users we can re-evaluate and allow a more flexible model

Reply:
Given you agree with my previous comment, this is something we likely want to do for both identifier and pattern mode?